src/mongocxx/test/spec/unified_tests/assert.cpp

```diff
@@ -190,26 +208,34 @@ void matches_document(types::bson_value::view actual,
     CAPTURE(to_json(actual_doc));
 
     for (auto&& kvp : expected_doc) {
-        match_scope_doc_key scope_key{string::to_string(kvp.key())};
+        std::string original_key = std::string(kvp.key());
+        std::string key = get_alternate_name(original_key);
+
         if (is_special(kvp)) {
-            if (!actual_doc[kvp.key()]) {
+            if (!actual_doc[key]) {
                 special_operator(value(nullptr), kvp.get_document(), map, false);
                 continue;
             }
         }
 
-        if (!actual_doc[kvp.key()]) {
-            FAIL("expected field '" + string::to_string(kvp.key()) +
-                 "' to be present, but it is absent");
+        if (auto n_inserted = actual_doc["nInserted"]) {
+            auto num_inserted = n_inserted.get_int32();
+            // If nothing was inserted, then we don't get "upsertedIds", so skip
+            // the check for it.
+            if ((num_inserted == 0) && (key == "upsertedIds")) {
+                continue;
+            }
         }
 
-        assert::matches(actual_doc[kvp.key()].get_value(), kvp.get_value(), map, false);
-        --extra_fields;
-    }
+        if (actual_doc[key]) {
+            assert::matches(actual_doc[key].get_value(), kvp.get_value(), map, false);
+        } else if (actual_doc[original_key]) {
+            assert::matches(actual_doc[original_key].get_value(), kvp.get_value(), map, false);
```

Comment:
```suggestion
        if (const auto doc = actual_doc[key]) {
            assert::matches(doc.get_value(), kvp.get_value(), map, false);
        } else if (const auto doc = actual_doc[original_key]) {
            assert::matches(doc.get_value(), kvp.get_value(), map, false);
```
Avoid repeated field lookups.